### PR TITLE
fix(interp): send interrupt/kill signals to the whole process group

### LINKED ISSUE #1171: interp: killing a process does not always kill its children processes
This is something that affects Task, particularly in its watch funcionality, and also other of our projects that uses the interpreter.

It's mostly commonly perceived when running Go programs with `go run`, because the program is a child and it won't be killed when the user expects it to.

This is something that I noticed for a while, but I'm opening the issue only now. I plan to debug and attempt a fix, but it'd be useful to know if you have any direction to how to potentially address this.

I _think_ the problem doesn't happen on Windows, but I'm yet to test it.

I created a repository with the reproduction: https://github.com/andreynering/mvdan-sh-kill-issue-reproduction

### PR DESCRIPTION
**Pull request by andreynering**
* Fixes #1171

Changes:

* On Unix platforms, asking Go to set the process group ID to the child
* When sending an interrupt or kill signal, send it to the whole process group
* On `gosh`, ensure that we capture the signals and give time for the interpreter to kill its processes if needed

### CODE REVIEW

**Comment by andreynering** [comment]
I'll take a look at the CI failures and also do some manual testing on Linux and Windows to ensure it's working fine.

**Comment by andreynering** [comment]
CI is fixed.

I just did some testing on Windows and the veredict is: it's certainly improved, just not 100% because on some edge cases the child process is still not killed.

My opinion is that we could merge this as is as it's for sure an improvement, and I can always return later to improve it further for Windows.

**Review by mvdan — COMMENTED** [review]
A test would be nice, because hanging grandchildren is a tricky edge case. I can assist with that if you're stuck.

**Comment by mvdan** [comment]
Yes - in theory writing such a test with children and grandchildren is possible, via e.g. `TestMain` in the test executable behaving as different main programs depending on an input like an environment variable.

However it is quite a bit of added work and test machinery, and the logic here is rather simple, so I think it's OK to omit for now. I'll maybe look into it in the future if this logic gets more involved.

File: cmd/gosh/main.go [hunk 1]
```diff
@@ -100,3 +100,27 @@ func runInteractive(r *interp.Runner, stdin io.Reader, stdout, stderr io.Writer)
 	}
 	return runErr
 }
+
+func getContext() context.Context {
+	const maxSignals = 3
```
[the remark this answers is not in the document]
Comment (mvdan) [reply cmd/gosh/main.go]: understood. AFAICT this should only matter if `interp` itself does not properly exit or kill its children, but its code when running external programs does a proper kill signal if they don't respond to a polite interrupt signal, so external programs should not cause any issues. If interp itself is the one hanging, then that's a bug I want to be aware of and fix.
